climada/hazard/base.py

```diff
@@ -1469,63 +1519,10 @@ def concat(haz_list):
-        # check type and unit consistency among hazards
-        haz_types = {haz.tag.haz_type for haz in haz_list if haz.tag.haz_type != ''}
-        if len(haz_types) > 1:
-            raise ValueError("haz_list contains hazards of different "
-                            f"types: {haz_types}. The hazards are incompatible "
-                            "and cannot be concatenated.")
-
-        haz_types = {type(haz) for haz in haz_list}
-        if len(haz_types) > 1:
-            raise ValueError("haz_list contains hazards of different "
-                            f"types: {haz_types}. The hazards are incompatible "
-                            "and cannot be concatenated.")
-
-        units = {haz.units for haz in haz_list if haz.units != ''}
-        if len(units) > 1:
-            raise ValueError("haz_list contains hazards with different "
-                            f"units {units}. The hazards are incompatible and "
-                            "cannot be concatenated.")
-        elif len(units) == 0:
-            units = {''}
-
-        centroids = Centroids.union(*[haz.centroids for haz in haz_list])
         haz_concat = haz_list[0].__class__()
-        haz_concat.units = units.pop()
-        haz_concat.centroids = centroids
-
-        # indices for mapping matrices onto common centroids
-        hazcent_in_cent_idx_list = [
-            u_coord.assign_coordinates(haz.centroids.coord, centroids.coord,
-                                       threshold=0)
-            for haz in haz_list
-            ]
-
-        # concatenate attributes - hazards are assumed to have the same attributes
-        for attr_name in vars(haz_list[0]).keys():
-            attr_val_list = [getattr(haz, attr_name) for haz in haz_list]
-            if isinstance(attr_val_list[0], sparse.csr.csr_matrix):
-                # map sparse matrix onto centroids.
-                matrix = (
-                    sparse.csr_matrix(
-                        (matrix.data, cent_idx[matrix.indices], matrix.indptr),
-                        shape=(matrix.shape[0], centroids.size)
-                        )
-                    for matrix, cent_idx in zip(attr_val_list, hazcent_in_cent_idx_list)
-                    )
-                setattr(haz_concat, attr_name, sparse.vstack(matrix, format='csr'))
-            elif (isinstance(attr_val_list[0], np.ndarray)
-                  and attr_val_list[0].ndim == 1):
-                setattr(haz_concat, attr_name, np.hstack(attr_val_list))
-            elif isinstance(attr_val_list[0], list):
-                setattr(haz_concat, attr_name, sum(attr_val_list, []))
-            elif isinstance(attr_val_list[0], TagHazard):
-                for tag in attr_val_list:
-                    if tag is not haz_concat.tag: haz_concat.tag.append(tag)
-
-        haz_concat.sanitize_event_ids()
-
+        haz_concat.tag.haz_type = haz_list[0].tag.haz_type
+        haz_concat.units = haz_list[0].units
+        haz_concat.append(*haz_list)
```

Comment:
By delegating the static concat-operation to the non-static append-operation, we avoid the use of `__dict__`.

Reply:
But why not just make a copy of the first element in list, and then append the rest?

```
haz_concat = copy.deepcopy(haz_list[0])
haz_concat.append(*haz_list[1:])
```

Reply:
In the extreme case where the first element is an extremely large Hazard, this would copy the large Hazard's data in memory here and then copy it again when concatenating the attributes so that there is a moment in the execution of the code where you have three times the memory requirements of that Hazard, instead of just twice. This sounds quite hypothetical, but I run into out-of-memory problems all the time because of all kinds of reasons, so I wouldn't be surprised to see this scenario at some point.

Reply:
However, this will erase all the non-list attributes, which would be a different behaviour from `append`. This is not necessarily surprising, still noteworthy.

Reply:
Okay, then I propose to copy all non-list attributes:
https://github.com/CLIMADA-project/climada_python/blob/5bc9eb4155cf662988e7ce54a06ff9de4faedbcc/climada/hazard/base.py#L1542-L1546

Reply:
I just made a suggestion to make it very clear that the 'not simple' attributes are explicitly handled by `Hazard.append`.